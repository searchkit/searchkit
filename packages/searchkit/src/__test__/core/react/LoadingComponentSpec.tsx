import * as React from "react";

import {
  SearchkitManager,
  LoadingComponent
} from "../../../"

import {mount} from "enzyme";

describe("SearchkitProvider", ()=> {

  beforeEach(()=> {
    this.searchkit = SearchkitManager.mock()
    this.wrapper = mount(
      <LoadingComponent searchkit={this.searchkit}>
        <p>loading...</p>
      </LoadingComponent>
    )
  })

  it("display nothing when not loading", ()=> {
    expect(this.wrapper).toMatchSnapshot()
  })

  it("display children when loading", ()=> {
    this.searchkit.loading = true
    this.wrapper.instance().forceUpdate()
    expect(this.wrapper).toMatchSnapshot()
  })


})
