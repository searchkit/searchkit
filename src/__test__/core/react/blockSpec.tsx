import {
  block
} from "../../../"

describe("block", ()=> {

  it("block test", ()=> {
    let b = block("foo")
    // console.log(typeof b)
    // console.log(b.toString())
    // console.log(b.mix("bar").state({hover:true}).toString())
    let bemBlocks = {
      item:block("item"),
      container:block("container")
    }

    expect(bemBlocks.container.el("panel").mix("red").state({active:true}).toString())
      .toEqual("container__panel red is-active")

    expect(bemBlocks.container.el("panel").mix(bemBlocks.item.el().mod({mod:"yes"})).state({active:true}).toString())
      .toEqual("container__panel item item_mod_yes is-active")

    expect(bemBlocks.container.mod({type:"text"}).mix("red").state({active:true}).toString())
      .toEqual("container container_type_text red is-active")

    expect(block("button").el("icon").mod({name:"check"}).toString())
      .toEqual("button__icon button__icon_name_check")

    expect(block("button").el().mod({name:"check"}).toString())
      .toEqual("button button_name_check")


    expect(block("button").el().mix(["foo", "bar"]).toString())
      .toEqual("button foo bar")

  })

})
