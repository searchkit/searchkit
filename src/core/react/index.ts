export * from "./SearchkitComponent"
export * from "./SearchkitProvider"
export * from "./LoadingComponent"
export * from "./ReactComponentType"
export * from "./FastClick"
export * from "./pure-render/index"
