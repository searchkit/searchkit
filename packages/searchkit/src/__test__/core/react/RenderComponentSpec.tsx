import * as React from "react";
import * as PropTypes from "prop-types";
import {mount} from "enzyme";
import {
  fastClick
} from "../../../components/__test__/TestHelpers"

import {Panel} from "../../../components"

import {
  renderComponent,
  RenderComponentType,
  RenderComponentPropType
} from "../../../"


describe("RenderComponent", ()=> {

  beforeEach(()=> {
    this.SubPanel = class SubPanel extends Panel {
    }
    this.SubPanel.defaultProps.title = "SubPanel"
    this.SubPanelElement = <Panel title="PanelElement"/>

    this.PanelReactClass = class PanelReactClass extends React.Component {
      static contextTypes = {
        color:PropTypes.string
      }
      render(){
        return (<Panel title={"PanelReactClass " + this.context.color} {...this.props}>
          {this.props.children}
        </Panel>)
      }
    }



    this.PanelFunction = (props, context)=> {
      return (
        <Panel title={"PanelFunction " + context.color}>
          {props.children}
        </Panel>
      )
    }
    this.PanelFunction.contextTypes = {
      color:PropTypes.string
    }

    class Provider extends React.Component<any, any>{
      static childContextTypes = {
        color: PropTypes.string
      }
      getChildContext(){
        return {color:"purple"}
      }
      render(){
        return this.props.children
      }
    }

    this.mount = (component, props={})=> {
      this.wrapper = mount(
        <Provider>
          {renderComponent(
            component, props,
            <p>content..</p>
          )}
        </Provider>
      )
    }

  })

  it("React.Component class", ()=> {
    this.mount(this.SubPanel)
    expect(this.wrapper).toMatchSnapshot()

  })

  it("React Element", ()=> {
    this.mount(this.SubPanelElement)
    expect(this.wrapper).toMatchSnapshot()
  })

  it("React Class", ()=> {
    this.mount(this.PanelReactClass)
    expect(this.wrapper).toMatchSnapshot()
  })

  it("Render function", ()=> {
    this.mount(this.PanelFunction)
    expect(this.wrapper).toMatchSnapshot()
  })

  it("Invalid component", ()=> {
    spyOn(console, "warn")
    try{
      this.mount(10)
    } catch (e){

    }
    expect(console.warn).toHaveBeenCalledWith(
      "Invalid component", 10
    )

  })

})
