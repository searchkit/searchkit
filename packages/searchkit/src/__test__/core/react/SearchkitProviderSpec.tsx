import * as React from "react";

import {
  SearchkitProvider,
  SearchkitManager,
  SearchkitComponent
} from "../../../"

import {mount} from "enzyme";

describe("SearchkitProvider", ()=> {

  beforeEach(()=> {
    this.searchkit = SearchkitManager.mock()
    class SomeComponent extends SearchkitComponent<any, any> {

      render(){
        return <h1>Hello</h1>
      }
    }
    this.wrapper = mount(
      <SearchkitProvider searchkit={this.searchkit}>
        <SomeComponent/>
      </SearchkitProvider>
    )
  })

  it("searchkit provider should work correctly", ()=> {
    expect(this.wrapper.html()).toBe("<h1>Hello</h1>")
    expect(this.wrapper.node.props.searchkit)
      .toBe(this.searchkit)
  })

  it("should call setupListeners()", ()=> {
    spyOn(this.searchkit, "setupListeners")
    expect(this.searchkit.setupListeners).not.toHaveBeenCalled()
    this.wrapper.node.componentWillMount()
    expect(this.searchkit.setupListeners).toHaveBeenCalled()
    this.searchkit.guidGenerator.counter = 10
    this.searchkit.unlistenHistory = jasmine.createSpy("unlisten")
    this.wrapper.node.componentWillUnmount()
    expect(this.searchkit.unlistenHistory).toHaveBeenCalled()
    expect(this.searchkit.guidGenerator.counter).toEqual(0)

  })


})
