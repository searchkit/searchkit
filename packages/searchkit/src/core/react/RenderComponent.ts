import * as React from 'react'
import * as PropTypes from "prop-types"

const omitBy = require("lodash/omitBy")
const isUndefined = require("lodash/isUndefined")

export type RenderFunction = (props?:any, children?:any) => Element
export type Element = React.ReactElement<any>
export type RenderComponentType<P> = React.ComponentClass<P> | React.ClassicComponentClass<P> | Element | RenderFunction | any

export const RenderComponentPropType = PropTypes.oneOfType([
  function(props:any, propName: string) {
     if(isUndefined(props[propName]) || (props[propName]["prototype"] instanceof React.Component)) {
       return null
     }
  },
  PropTypes.element,
  PropTypes.func,
])

export function renderComponent(component:RenderComponentType<any>, props={}, children=null): React.ReactElement<any> {
  let isReactComponent = (
    component["prototype"] instanceof React.Component ||
    (component["prototype"] && component["prototype"].isReactComponent) ||
    typeof component === 'function'
  )
  if (isReactComponent){
    return React.createElement(
      component as React.ComponentClass<any>,
      props, children
    )
  } else if (React.isValidElement(component)){
    return React.cloneElement(
      component as Element,
      omitBy(props, isUndefined), children
    );
  }
  console.warn("Invalid component", component)
  return null
}
