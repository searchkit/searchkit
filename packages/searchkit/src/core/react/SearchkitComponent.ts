import * as React from "react"
import * as PropTypes from "prop-types"
import {SearchkitManager} from "../SearchkitManager"
import {ImmutableQuery} from "../query"
import {Accessor} from "../accessors/Accessor"
import {Utils} from "../support"
const mapValues = require("lodash/mapValues")
import {block} from "./block"

export interface SearchkitComponentProps {
  mod?:string
  className?:string
  translations?:Object
  searchkit?:SearchkitManager
  key?:string
}

export class SearchkitComponent<P extends SearchkitComponentProps,S> extends React.Component<P,S> {
  _searchkit:SearchkitManager
  accessor:Accessor
  stateListenerUnsubscribe:Function
  translations:Object = {}
  unmounted = false

  static contextTypes: React.ValidationMap<any> = {
		searchkit:PropTypes.instanceOf(SearchkitManager)
	}

  static translationsPropType = (translations)=> {
    return PropTypes.shape(mapValues(translations, ()=> PropTypes.string))
  }

  static propTypes:any = {
    mod :PropTypes.string,
    className :PropTypes.string,
    translations: PropTypes.objectOf(
      PropTypes.string),
    searchkit:PropTypes.instanceOf(SearchkitManager)
  }

  constructor(props?){
    super(props)
    this.translate = this.translate.bind(this)
  }

  defineBEMBlocks() {
    return null;
  }

  defineAccessor():Accessor{
    return null
  }

  translate(key, interpolations?){
    let translation = (
      (this.searchkit.translate(key)) ||
      (this.props.translations && this.props.translations[key]) ||
      this.translations[key] || key)
    return Utils.translate(translation, interpolations)
  }

  get bemBlocks(): any {
    return mapValues(this.defineBEMBlocks(), (cssClass) => {
      return block(cssClass).el
    })
  }

  get searchkit(): SearchkitManager {
    return this._searchkit || (this._searchkit = this._getSearchkit());
  }

  _getSearchkit(): SearchkitManager {
    return this.props.searchkit || this.context["searchkit"];
  }

  set searchkit(value: SearchkitManager) {
    this._searchkit = value;
  }

  componentDidMount() {
    this._initAccessor();
    if (this.searchkit) {
      this.stateListenerUnsubscribe = this.searchkit.emitter.addListener(() => {
        if (!this.unmounted) {
          this.forceUpdate();
        }
      });
    }
    this.forceUpdate();
  }
  
  _initAccessor() {
    if (this.searchkit && !this.accessor) {
      this.accessor = this.defineAccessor();
      if (this.accessor) {
        this.accessor = this.searchkit.addAccessor(this.accessor);
        return true;
      }
    }
    if (!this.searchkit) {
      console.warn(
        "No searchkit found in props or context for " + this.constructor["name"]
      );
    }
    return false;
  }

  componentWillUnmount(){
    if(this.stateListenerUnsubscribe){
		  this.stateListenerUnsubscribe()
    }
    if(this.searchkit && this.accessor){
      this.searchkit.removeAccessor(this.accessor)
    }
    this.unmounted = true
	}

  getResults(){
    return this.searchkit.results
  }

  getHits(){
    return this.searchkit.getHits()
  }

  getHitsCount(){
    return this.searchkit.getHitsCount()
  }

  hasHits(){
    return this.searchkit.hasHits()
  }

  hasHitsChanged(){
    return this.searchkit.hasHitsChanged()
  }

  getQuery():ImmutableQuery {
    return this.searchkit.query
  }

  isInitialLoading(){
    return this.searchkit.initialLoading
  }

  isLoading(){
    return this.searchkit.loading
  }

  getError(){
    return this.searchkit.error
  }


}
