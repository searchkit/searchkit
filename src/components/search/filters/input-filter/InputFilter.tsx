import * as React from "react";

import {
  QueryAccessor,
  SearchkitComponent,
  SearchkitComponentProps,
  ReactComponentType,
  renderComponent
} from "../../../../core";

import { SearchBox } from "../../search-box/SearchBox"

import {
  Panel
} from "../../../ui"

import {defaults, throttle, assign, isUndefined} from 'lodash'

export interface InputFilterProps extends SearchkitComponentProps {
  id: string
  title: string
  mod?: string
  searchOnChange?:boolean
  searchThrottleTime?:number
  queryBuilder?:Function
  queryFields?:Array<string>
  queryOptions?:any
  prefixQueryFields?:Array<string>
  prefixQueryOptions?:any
  placeholder?: string
  blurAction?:"search"|"restore"
  containerComponent?: ReactComponentType<any>
}

export class InputFilter extends SearchkitComponent<InputFilterProps, any> {
  accessor:QueryAccessor
  lastSearchMs:number
  throttledSearch: () => void

  static translations:any = {
    "searchbox.placeholder":"Search"
  }
  translations = SearchBox.translations

  static defaultProps = {
    containerComponent: Panel,
    collapsable: false,
    mod: "sk-input-filter",
    searchThrottleTime:200,
    blurAction: "search"
  }

  static propTypes = defaults({
    id:React.PropTypes.string.isRequired,
    title: React.PropTypes.string.isRequired,
    searchOnChange:React.PropTypes.bool,
    searchThrottleTime:React.PropTypes.number,
    queryBuilder:React.PropTypes.func,
    queryFields:React.PropTypes.arrayOf(React.PropTypes.string),
    queryOptions:React.PropTypes.object,
    prefixQueryFields:React.PropTypes.arrayOf(React.PropTypes.string),
    prefixQueryOptions:React.PropTypes.object,
    translations:SearchkitComponent.translationsPropType(
      SearchBox.translations
    ),
    mod: React.PropTypes.string,
    placeholder: React.PropTypes.string,
    blurAction: React.PropTypes.string
  }, SearchkitComponent.propTypes)

  constructor (props:InputFilterProps) {
    super(props);
    this.state = {
      focused:false,
      input: undefined
    }
    this.lastSearchMs = 0
    this.onClear = this.onClear.bind(this)
    this.throttledSearch = throttle(()=> {
      this.searchQuery(this.accessor.getQueryString())
    }, props.searchThrottleTime)
  }

  componentWillMount() {
    super.componentWillMount()
  }

  defineBEMBlocks() {
    return { container:this.props.mod };
  }

  defineAccessor(){
    const {
      id, title, prefixQueryFields, queryFields, queryBuilder,
      searchOnChange, queryOptions, prefixQueryOptions } = this.props
    return new QueryAccessor(id, {
      title,
      addToFilters: true,
      queryFields:queryFields || ["_all"],
      prefixQueryFields,
      queryOptions:assign({}, queryOptions),
      prefixQueryOptions:assign({}, prefixQueryOptions),
      queryBuilder,
      onQueryStateChange: () => {
        if (!this.unmounted && this.state.input){
          this.setState({input: undefined})
        }
      }
    })
  }

  onSubmit(event) {
    event.preventDefault()
    this.searchQuery(this.getValue())
  }

  searchQuery(query) {
    let shouldResetOtherState = false
    this.accessor.setQueryString(query, shouldResetOtherState )
    let now = +new Date
    let newSearch = now - this.lastSearchMs <= 2000
    this.lastSearchMs = now
    this.searchkit.performSearch(newSearch)
  }

  getValue(){
    const { input } = this.state
    if (isUndefined(input)) {
      return this.getAccessorValue()
    } else {
      return input
    }
  }

  getAccessorValue(){
    return (this.accessor.state.getValue() || "") + ""
  }

  onChange(e){
    const query = e.target.value;
    if (this.props.searchOnChange) {
      this.accessor.setQueryString(query)
      this.throttledSearch()
      this.forceUpdate()
    } else {
      this.setState({ input: query })
    }
  }

  onClear(){
    this.accessor.state = this.accessor.state.clear()
    this.searchkit.performSearch()
    this.setState({ input: undefined })
  }

  setFocusState(focused:boolean) {
    if (!focused){
      const { input } = this.state
      if (this.props.blurAction == "search"
        && !isUndefined(input)
        && input != this.getAccessorValue()){
        this.searchQuery(input)
      }
      this.setState({
        focused,
        input: undefined // Flush (should use accessor's state now)
      })
    } else {
      this.setState({ focused })
    }
  }

  render() {
    const { containerComponent, title, id } = this.props
    const block = this.bemBlocks.container
    const value = this.getValue()
    return renderComponent(containerComponent, {
      title,
      className: id ? `filter--${id}` : undefined,
      disabled: (this.searchkit.getHitsCount() == 0) && (this.getAccessorValue() == "")
    },
      <div className={block().state({focused:this.state.focused})}>
        <form onSubmit={this.onSubmit.bind(this)}>
          <div className={block("icon")}></div>
          <input type="text"
            data-qa="input-filter"
            className={block("text")}
            placeholder={this.props.placeholder || this.translate("searchbox.placeholder")}
            value={value}
            onFocus={this.setFocusState.bind(this, true)}
            onBlur={this.setFocusState.bind(this, false)}
            ref="queryField"
            autoFocus={false}
            onInput={this.onChange.bind(this)}/>
          <input type="submit" value="search" className={block("action")} data-qa="submit"/>
          <div data-qa="remove"
               onClick={this.onClear}
               className={block("remove").state({hidden:value == ""})} />
        </form>
      </div>
    );
  }

}
