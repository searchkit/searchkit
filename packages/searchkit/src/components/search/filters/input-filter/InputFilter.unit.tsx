import * as React from "react";
import {mount} from "enzyme";
import { InputFilter } from "./InputFilter";
import { 
  SearchkitManager, SimpleQueryString, QueryString, QueryAccessor
} from "../../../../core";

import {
  fastClick
} from "../../../__test__/TestHelpers"

import {
  Panel
} from "../../../ui"

import * as sinon from "sinon";

const omit = require("lodash/omit")

describe("InputFilter tests", () => {

  beforeEach(() => {

    this.searchkit = SearchkitManager.mock()
    spyOn(this.searchkit, "performSearch")

    this.searchkit.translateFunction = (key)=> {
      return {
        "searchbox.placeholder":"search placeholder",
        "searchbox.button":"submit"
      }[key]
    }

    this.createWrapper = (searchOnChange=false, queryFields=null, prefixQueryFields=null, otherProps={}) => {
      this.wrapper = mount(
        <InputFilter searchkit={this.searchkit}
                   id="test_id"
                   title="Test title"
                   searchOnChange={searchOnChange}
                   queryFields={queryFields}
                   prefixQueryFields={prefixQueryFields}
                   {...otherProps} />
      );
      this.accessor = this.searchkit.getAccessorByType(QueryAccessor)
    }

    this.setResults = ()=> {
      this.searchkit.setResults({
        hits:{
          hits:[{_id:1, title:1},{_id:2,title:2}],
          total:2
        }
      })
    }

    this.setEmptyResults = () => {
      this.searchkit.setResults({
        hits: {
          total: 0
        }
      })
    }

    this.typeSearch = (value)=> {
      this.wrapper.find(".sk-input-filter__text")
        .simulate("input", {target:{value}})
    }

  });

  it("render", () => {
    this.createWrapper()
    expect(this.wrapper).toMatchSnapshot()
  })

  it("toggles visibility", () => {
    let spy = sinon.spy()
    this.searchkit.performSearch = spy
    this.createWrapper(true)

    this.setEmptyResults()
    this.wrapper = this.wrapper.update()
    expect(this.wrapper.find(".sk-panel").hasClass("is-disabled")).toBe(true)

    this.setResults()
    this.wrapper = this.wrapper.update()
    expect(this.wrapper.find(".sk-panel").hasClass("is-disabled")).toBe(false)

    // Don't hide if active filter
    this.typeSearch("noresults")
    this.setEmptyResults()
    expect(this.wrapper.find(".sk-panel").hasClass("is-disabled")).toBe(false)
  })

  it("should allow custom mod and className", () => {
    this.createWrapper(false, null, null, {
      mod: "my-input", className: "my-class"
    })
    this.setResults()
    expect(this.wrapper).toMatchSnapshot()
  })

  it("search on change", () => {
    let spy = sinon.spy()
    this.searchkit.performSearch = spy
    this.createWrapper(true)
    this.typeSearch("m")
    expect(this.accessor.state.getValue()).toBe("m")
    expect(spy.callCount).toBe(1)
    this.typeSearch("ma")
    expect(this.accessor.state.getValue()).toBe("ma")
    expect(spy.callCount).toBe(1) // throttling should block it
    this.wrapper.instance().throttledSearch.flush()
    expect(spy.callCount).toBe(2)
  })

  it("search on change with clock", ()=> {
    let queries = []
    this.searchkit.performSearch = ()=> {
      queries.push(this.searchkit.buildQuery())
    }
    this.createWrapper(true)
    expect(this.wrapper.instance().props.searchThrottleTime).toBe(200)
    this.typeSearch("m")
    this.wrapper.instance().throttledSearch.flush()
    expect(queries.length).toBe(1)
    expect(queries[0].getSelectedFilters()[0].value).toBe("m")
    this.typeSearch("ma")
    expect(queries.length).toBe(1)
    this.wrapper.instance().throttledSearch.flush()
    expect(queries.length).toBe(2)
    expect(queries[1].getSelectedFilters()[0].value).toBe("ma")
  })

  it("search on submit", () => {
    let spy = sinon.spy()
    this.searchkit.performSearch = spy

    this.createWrapper(false)
    this.typeSearch('m')
    this.typeSearch('ma')
    // State left in the component
    expect(this.accessor.state.getValue()).toBe(null)
    expect(spy.callCount).toBe(0)
    this.wrapper.find("form").simulate("submit")
    expect(this.accessor.state.getValue()).toBe("ma")
    expect(spy.callCount).toBe(1)
  })

  it("should have a working remove icon", () => {
    let spy = sinon.spy()
    this.searchkit.performSearch = spy

    this.createWrapper(false)
    this.setResults()

    expect(this.wrapper.find(".sk-input-filter__remove").hasClass("is-hidden")).toBe(true)
    this.typeSearch('ma')
    expect(this.wrapper.find(".sk-input-filter__remove").hasClass("is-hidden")).toBe(false)
    expect(spy.callCount).toBe(0)
    this.wrapper.find("form").simulate("submit")
    expect(spy.callCount).toBe(1)
    this.wrapper.find(".sk-input-filter__remove").simulate("click")
    expect(this.accessor.state.getValue()).toBe(null)
    expect(spy.callCount).toBe(2)
    expect(this.wrapper.find(".sk-input-filter__remove").hasClass("is-hidden")).toBe(true)
  })

  it("should configure accessor defaults correctly", ()=> {
    this.createWrapper(false, ["title"])

    expect(this.accessor.key).toBe("test_id")
    let options = this.accessor.options
    expect(options).toEqual({
      title: "Test title",
      addToFilters: true,
      queryFields: ["title"],
      prefixQueryFields:null,
      queryOptions: {},
      prefixQueryOptions: {},
      queryBuilder:undefined,
      onQueryStateChange:jasmine.any(Function)
    })

  })

  it("should configure accessor search on change correctly", ()=> {
    this.createWrapper(true, ["title"])

    expect(this.accessor.key).toBe("test_id")
    let options = this.accessor.options
    expect(options).toEqual({
      title: "Test title",
      addToFilters: true,
      prefixQueryFields:null,
      queryFields: ["title"],
      queryOptions: {},
      prefixQueryOptions: {},
      queryBuilder:undefined,
      onQueryStateChange:jasmine.any(Function)
    })

  })

  it("should configure accessor + prefix", ()=> {
    this.createWrapper(true, ["title"], ["prefix"], {
      queryOptions:{minimum_should_match:"60%"},
      prefixQueryOptions:{minimum_should_match:"70%"},
      queryBuilder:QueryString
    })

    expect(this.accessor.key).toBe("test_id")
    let options = this.accessor.options
    expect(options).toEqual({
      title: "Test title",
      addToFilters: true,
      queryFields: ["title"],
      prefixQueryFields:["prefix"],
      queryOptions:{minimum_should_match:"60%"},
      prefixQueryOptions:{minimum_should_match:"70%"},
      queryBuilder:QueryString,
      onQueryStateChange:jasmine.any(Function)
    })
  })

  it("should accept Panel elements as containerComponent", ()=> {
    this.createWrapper(true, ["title"], ["prefix"], {
      containerComponent: <Panel collapsable={true} />
    })
    expect(this.wrapper.find(".sk-panel__header").hasClass("is-collapsable")).toBe(true)
  })

  describe("url change + blurAction", ()=> {

    it("blurAction:restore", ()=> {
      this.createWrapper(false, ["title"], ["prefix"], {
        blurAction:"restore"
      })
      this.typeSearch("la")
      expect(this.wrapper.instance().getValue() ).toEqual("la")
      this.accessor.fromQueryObject({
        test_id:"foo"
      })
      expect(this.wrapper.instance().getValue() ).toEqual("foo")

      this.typeSearch("bar")
      expect(this.wrapper.instance().getValue()).toEqual("bar")
      this.wrapper.find(".sk-input-filter__text")
        .simulate("blur")

      // should be restored to previous value
      expect(this.wrapper.instance().getValue()).toEqual("foo")
      expect(this.searchkit.performSearch).not.toHaveBeenCalled()

    })

    it("blurAction:search", ()=> {
      this.createWrapper(false, ["title"], ["prefix"], {
        blurAction:"search"
      })
      this.typeSearch("la")
      expect(this.wrapper.instance().getValue() ).toEqual("la")
      this.accessor.fromQueryObject({
        test_id:"foo"
      })
      expect(this.wrapper.instance().getValue() ).toEqual("foo")

      this.typeSearch("bar")
      expect(this.wrapper.instance().getValue()).toEqual("bar")
      this.wrapper.find(".sk-input-filter__text")
        .simulate("blur")

      // should flush value + search
      expect(this.wrapper.instance().getValue()).toEqual("bar")
      expect(this.searchkit.performSearch).toHaveBeenCalled()

    })

  })


});
