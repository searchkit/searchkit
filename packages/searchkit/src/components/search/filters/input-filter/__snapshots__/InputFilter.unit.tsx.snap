// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputFilter tests render 1`] = `
<div class="sk-panel filter--test_id is-disabled">
  <div class="sk-panel__header is-disabled">Test title</div>
  <div class="sk-panel__content is-disabled">
    <div class="sk-input-filter">
      <form>
        <div class="sk-input-filter__icon"></div><input type="text" data-qa="input-filter" class="sk-input-filter__text" placeholder="search placeholder" value=""><input type="submit" class="sk-input-filter__action" data-qa="submit" value="submit">
        <div data-qa="remove" class="sk-input-filter__remove is-hidden"></div>
      </form>
    </div>
  </div>
</div>
`;

exports[`InputFilter tests should allow custom mod and className 1`] = `
<div class="sk-panel filter--test_id">
  <div class="sk-panel__header">Test title</div>
  <div class="sk-panel__content">
    <div class="my-input">
      <form>
        <div class="my-input__icon"></div><input type="text" data-qa="input-filter" class="my-input__text" placeholder="search placeholder" value=""><input type="submit" class="my-input__action" data-qa="submit" value="submit">
        <div data-qa="remove" class="my-input__remove is-hidden"></div>
      </form>
    </div>
  </div>
</div>
`;
